lib/paper_trail/serializer.ex

```diff
@@ -178,25 +175,23 @@ defmodule PaperTrail.Serializer do
     {field, serialize(value, options, event)}
   end
 
+  defp dump_field!({field, [%Ecto.Changeset{} | _] = changesets}, _schema, _adapter, options, event) do
```

Comment:
can you clarify this new `dump_field!` clause and the removal of `do_dump_field!`? Are they related?

Reply:
Yes, they are related. The `do_dump_field!` was a quick workaround, but it introduced a bug (the tests caught):

```elixir
  defp do_dump_field!(schema, _field, {_, Ecto.Embedded, _}, _value, _adapter) do
    Ecto.embedded_dump(schema.__struct__(), :json)
  end
```

Notice that the `schema.__struct__()` is wrong. First because the `schema` we receive here is not the embedded schema, its the "main" schema. The second problem is that calling `__struct__()` will create a struct with nil values, thus, not serializing any changes. For example, in the tests we have a `Company` with a `EmailOptions` embedded schema. But the tests are failing because the serialized company version was wrong:

![image](https://user-images.githubusercontent.com/7737375/150988858-c3ca4708-5fd5-41b2-a2ca-544ed1fb80e5.png)

Basically this workaround was a fix for when we use `embeds_many` with `cast_embed` or `put_embed` and in the changes of the "main" changeset, there's a list of changesets for the embeds, but this case we need to handle "manually", otherwise this list of changesets is passed to the `Ecto.Type.adapter_dump(adapter, type, value)`, which raises an error since it cannot serialize a changesets.

Reply:
Got it, thanks!